main.go

```diff
@@ -280,54 +278,69 @@ func (s *Session) firstNav(ctx context.Context) error {
 	return nil
 }
 
-// navToEnd waits for the page to be ready to receive scroll key events, by
-// trying to select an item with the right arrow key, and then scrolls down to the
-// end of the page, i.e. to the oldest items.
+// navToEnd waits for the page to be ready to receive scroll key events,
+// and then scrolls down to the end of the page, i.e. to the oldest items.
 func navToEnd(ctx context.Context) error {
-	// wait for page to be loaded, i.e. that we can make an element active by using
-	// the right arrow key.
-	for {
-		chromedp.KeyEvent(kb.ArrowRight).Do(ctx)
-		time.Sleep(tick)
-		var ids []cdp.NodeID
-		if err := chromedp.Run(ctx,
-			chromedp.NodeIDs(`document.activeElement`, &ids, chromedp.ByJSPath)); err != nil {
-			return err
-		}
-		if len(ids) > 0 {
-			if *verboseFlag {
-				log.Printf("We are ready, because element %v is selected", ids[0])
-			}
-			break
-		}
-		time.Sleep(tick)
-	}
+	// wait for page to be loaded
+	chromedp.WaitReady("body", chromedp.ByQuery).Do(ctx)
 
-	// try jumping to the end of the page. detect we are there and have stopped
-	// moving when two consecutive screenshots are identical.
-	var previousScr, scr []byte
+	// Try scrolling to the end of the page.
+	// After each scroll attempt we extract the last Photo Page Element (href) from the DOM.
+	// We detect we are at the end when two consecutive DOM extractions are identical.
+	var previousHref string
 	for {
 		chromedp.KeyEvent(kb.PageDown).Do(ctx)
 		chromedp.KeyEvent(kb.End).Do(ctx)
-		chromedp.CaptureScreenshot(&scr).Do(ctx)
-		if previousScr == nil {
-			previousScr = scr
-			continue
+		time.Sleep(tick) // sleep *before* we establish new state in DOM
+
+		// Extract last Photo Page Element (href) from DOM.
+		lastHrefInDOM, err := lastPhotoInDOM(ctx)
+		if err != nil {
+			continue // Just ignore this error, continue will retry.
 		}
-		if bytes.Equal(previousScr, scr) {
+		if previousHref == lastHrefInDOM {
 			break
 		}
-		previousScr = scr
-		time.Sleep(tick)
+		previousHref = lastHrefInDOM
 	}
 
-	if *verboseFlag {
-		log.Printf("Successfully jumped to the end")
+	// Now that we have stopped scrolling, select (focus) on the last element
+	// The element must be focused, so that navToLast can send "\n" to enter photo detail page
+	lastEltSel := fmt.Sprintf(`a[href="%s"]`, previousHref)
+	if err := chromedp.Focus(lastEltSel).Do(ctx); err != nil {
+		return err
```

Comment:
I've just realized; doesn't this actually do the job of navToLast (except the actual navigation itself)? So couldn't we move that code to inside navToLast, so that it replaces there the loop that relies on the location change? And then we would only have to do one "\n" key event right after?

Alternatively, we could keep it here, and remove most of the the code in navToLast (it would just do the one "\n" event basically, but I don't like this solution as much as it does not separate as well each task.

Reply:
Yes, we could structure it that way.
but I also prefer keeping the tasks separate.

In my other experiment, I am trying to abstract the following:
- positioning: AlbumPage[First|Last],DetailPage[FirstLast]
- iterating: AlbumPage[Forward|Reverse], DetailPage[Forward|Reverse]
- itemCallback(item): [list,files,perkeep] 
  - *files and perkeep also have an optimistic variant which checks for existence before performing download*
- This way I can use the same outer loop for all combinations
- The iterators are using browser events [focuschanged,targetchanged] instead of polling and arbitrary `sleeps`. There is also an event fired for an initiated download, which obviates the need for the directory polling.

When I'm done finding the right abstractions I can port it back to Go/cdp.

Reply:
> Yes, we could structure it that way.
> but I also prefer keeping the tasks separate.

Hmm, I'm not sure we understood each other, so just to be clear:
1) do you agree that most of what navToLast is doing is redundant with and already done by L309 and L310?
2) if yes, do you agree to move L309 and L310 to inside navToLast?

If I'm not making sense, I can add myself a commit on top of this PR so you can see what I mean?

> In my other experiment, I am trying to abstract the following:
> 
> * positioning: AlbumPage[First|Last],DetailPage[FirstLast]
> * iterating: AlbumPage[Forward|Reverse], DetailPage[Forward|Reverse]
> * itemCallback(item): [list,files,perkeep]
>   
>   * _files and perkeep also have an optimistic variant which checks for existence before performing download_
> * This way I can use the same outer loop for all combinations
> * The iterators are using browser events [focuschanged,targetchanged] instead of polling and arbitrary `sleeps`. There is also an event fired for an initiated download, which obviates the need for the directory polling.
> 
> When I'm done finding the right abstractions I can port it back to Go/cdp.

Reply:
Thanks for clarifying.
No, *I don't agree that most of what navToLast is doing is redundant*

We still need `navToLast` to work the way it does, because the selected element (309-310), is not guaranteed to be the *actual* last element, it is just *close* to the end.

Reply:
Ok, understood, let's leave L309 and navToLast as they are for now.
Do you know why L309-L310 isn't enough to get the very last element though?